File: Symulator Świata/animal.py
import random
from organism import Organism
from coordinates import Coordinates


class Animal(Organism):

    def move(self, coord, speed):
        x = coord.x
        y = coord.y
        newCoord = Coordinates(x, y)
        dir = random.randint(0, 3)
        if dir == 0:
            newCoord.y -= speed
            if newCoord.y < 0:
                newCoord.y += speed * 2

        elif dir == 1:
            newCoord.x += speed
            if newCoord.x >= self.world.getColumns():
                newCoord.x -= speed * 2
        elif dir == 2:
            newCoord.y += speed
            if newCoord.y >= self.world.getRows():
                newCoord.y -= speed * 2
        elif dir == 3:
            newCoord.x -= speed
            if newCoord.x < 0:
                newCoord.x += speed * 2
        return newCoord

    def action(self, coord):
        newCoord = self.move(coord, 1)
        if newCoord != coord:
            self.collision(coord, newCoord)

    def collision(self, coord, newCoord):
        if newCoord in self.world.board.keys():  # jesli napotkano inny organizm
            if isinstance(self.world.board[newCoord], type(self)):  # sprawdzenie czy nastapi rozmnozenie
                if self.age != 0 and self.world.board[newCoord].age != 0:  # musi byc spelniony warunek wieku
                    self.multiply(newCoord)
            else:
                self.fight(coord, newCoord)
        else: # jesli nie napotkano innego organizmu
            self.world.board.pop(coord)
            self.world.board[newCoord] = self

    def fight(self, coord, newCoord):
        enemy = self.world.board[newCoord]
        if isinstance(self.world.board[newCoord], Animal):  # jeśli wrogim organizmem jest zwierzę to następuje walka
            if not enemy.ranAway(coord, newCoord):
                if not enemy.hasDeflectedAttack(coord, newCoord):
                    if self.strength >= enemy.strength:
                        self.world.commentator.addComment("{0} pokonal {1}".format(self.name, enemy.name))
                        self.world.board.pop(coord)
                        self.world.board[newCoord] = self
                    else:
                        self.world.commentator.addComment("{0} pokonal {1}".format(enemy.name, self.name))
                        self.world.board.pop(coord)
        else:  # jeśli wrogi organizm to roślina to jest ona zjadana i rozpatrywane są efekty jej zjedzenia
            if self.poisonResistant():  # odporność na truciznę
                enemy.hasIncreasedStrength(coord)
                self.world.commentator.addComment('{0} zjada {1}'.format(self.name, enemy.name))
                self.world.board[newCoord] = self.world.board[coord]
                self.world.board.pop(coord)
            elif not enemy.poisonous(coord, newCoord):
                enemy.hasIncreasedStrength(coord)
                self.world.commentator.addComment("{0} zjada {1}".format(self.name, enemy.name))
                self.world.board[newCoord] = self.world.board[coord]
                self.world.board.pop(coord)

File: PO PROJEKT 1/PO PROJEKT 1/PROJEKT PO PLIKI ŹRÓDŁOWE/Animal.cpp
#include "Animal.h"
void Animal::action(Coordinate coord) {
    Coordinate newCoord = move(coord, 1);
    if (newCoord == coord) {}
    else {
        collision(coord, newCoord);
    }
}

void Animal::collision(Coordinate coord, Coordinate newCoord) {
    if (world->board.count(newCoord)) {
        if (world->board[newCoord]->getSprite() == this->getSprite()) { 
            if (this->getAge() != 0 && world->board[newCoord]->getAge() != 0) { // jeœli wiek zwierz¹t nie jest równy 0
                multiply(newCoord);
            }
        }
        else {
            fight(coord, newCoord);
        }
    }
    else {
        world->board.erase(coord);
        world->board.insert({newCoord, this });
    }
}

void Animal::fight(Coordinate coord, Coordinate newCoord) { // pamiêæ po walce zwalniania jest w world.cpp w funkcji wykonuj¹cej turê
    if (dynamic_cast<Animal*>(world->board[newCoord])) { // gdy na nowym polu znajduje siê zwierze
        if (!world->board[newCoord]->ranAway(coord, newCoord) && !world->board[newCoord]->hasDeflectedAttack(coord, newCoord)) {
            if (world->board[coord]->getStrength() >= world->board[newCoord]->getStrength()) { 
                cout << "Inicjator to " << this->getSprite() << " ";
                cout << endl << "Walka- " << this->getSprite() << " pokonuje " << world->board[newCoord]->getSprite() << endl;
                world->board[newCoord] = world->board[coord];
                world->board.erase(coord);
            }
            else if (world->board[coord]->getStrength() < world->board[newCoord]->getStrength()) {
                cout << "inicjator to " << this->getSprite() << " ";
                cout << endl << "Walka- " << world->board[newCoord]->getSprite() << " pokonuje " << this->getSprite() << endl;
                world->board.erase(coord);
            }
        }
    }
    else { // jeœli zwierze natrafi na roœlinê zjada j¹
        cout << "Inicjator to " << this->getSprite() << " ";
        cout << endl << "Walka- " << this->getSprite() << " zjada " << world->board[newCoord]->getSprite() << endl;
        if (!world->board[newCoord]->poisonous(coord, newCoord)) {
            world->board[newCoord]->hasIncreasedStrength(coord);
            world->board[newCoord] = world->board[coord];   
            world->board.erase(coord);
        }
    }
}

Coordinate Animal::move(Coordinate coord, int speed) {
    Coordinate newCoord = coord;
    int dir = (rand() % 4);
    if (dir == N) {
        newCoord.y -= speed;
        if (newCoord.y < 0) {
            newCoord.y = 0;
        }
    }
    else if (dir == E) {
        newCoord.x += speed;
        if (newCoord.x >= world->getColumns()) {
            newCoord.x = world->getColumns() - 1;
        }
    }
    else if (dir == S) {
        newCoord.y += speed;
        if (newCoord.y >= world->getRows()) {
            newCoord.y = world->getRows() - 1;
        }
    }
    else if (dir == W) {
        newCoord.x -= speed;
        if (newCoord.x < 0) {
            newCoord.x = 0;
        }
    }
    return newCoord;
}
